### PR TITLE
Add configuration for data residency

### LINKED ISSUE #1680: Add another configuration map for data residency configuration
**Problem**
This configuration will be used to support #1678. But since the data residency itself is small, here are two possible ways:

1) create one for generic configuration in eventing so that new configurations can be added.
2) create one for data residency only.

**[Persona:](https://github.com/knative/eventing/blob/master/docs/personas.md)** 
Which persona is this feature for?

**Exit Criteria**
A new config map in `clound-run-events` namespace.

**Time Estimate (optional):** 
How many developer-days do you think this may take to resolve?

**Additional context (optional)**
Add any other context about the feature request here.

### PR DESCRIPTION
**Pull request by zhongduo**
Fixes #1680 

## Proposed Changes

- Add a new configuration map for data residency
- Update Topic CRD reconciler to use data residency configuration map
- Update broker and trigger reconciler to use data residency configuration when creating topics.

**Release Note**

```release-note
- 🎁 Add data residency support through configuration map.
```

**Docs**

### CODE REVIEW

**Comment by zhongduo** [comment]
This PR and the issue is open to input about if we should keep a config map called "config-data-residency" or make it larger scope to "config-gcp-features" which is what I am using now.

File: config/core/configmaps/data-residency.yaml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    # to users that `kubectl edit` this config map.
+    #
+    # These sample configuration options may be copied out of
+    # this example block and unindented to be in the data block
+    # to actually change the configuration.
+
+    # messagestoragepolicy.allowedpersistenceregions field specifies
+    # all the allowed regions for data residency. The default or an empty value will 
+    # mean no data residency requirement.
+    messagestoragepolicy.allowedpersistenceregions: "us-east1,us-west1"
```
[the remark this answers is not in the document]
Comment (zhongduo) [reply config/core/configmaps/data-residency.yaml:40]: Agreed. This field is a field used in the topic config, I was thinking to simply pass whatever in the field to the topic config, but I think the topic is also using `[]string`.

File: config/core/configmaps/data-residency.yaml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: config-data-residency
```
Comment (yolocs) [inline config/core/configmaps/data-residency.yaml:18]: I suggest rename this configmap to something like `pubsub-config`. I can imagine there will be other pubsub settings could fit here, e.g. CMEK.

File: config/core/configmaps/data-residency.yaml [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  namespace: cloud-run-events
+data:
+  _example: |
```
Comment (yolocs) [inline config/core/configmaps/data-residency.yaml:21]: I'm fine with creating such a placeholder since it does nothing. But I'm looking for more details on how this will be used.

1. How this configmap will be read? Volume mount? Watch?
2. Do you allow finer grained control? Like per broker? How it would impact the configmap structure?

**Review by yolocs — COMMENTED** [review]
Update the release note section in the description with something like "Add data residency configmap placeholder"

File: config/core/configmaps/data-residency.yaml (hunk 2 above)
Comment (zhongduo) [reply config/core/configmaps/data-residency.yaml:18]: The data residency requirements might be used other than pubsub in the future, though it seems that we are only using pubsub as the storage now. So I would say we should keep it.

File: config/core/configmaps/data-residency.yaml (hunk 3 above)
Comment (zhongduo) [reply config/core/configmaps/data-residency.yaml:21]: I will add the actual code for this configuration to this PR too. For your question:

1. This is still being discussed, but watch seems better since our system already uses watch for other config maps, also, it provides easier interface for reacting to the change.

2. Personally I think even if we allow per broker configuration, this will still serve as a default and same configuration will be added to per broker and get reconciled by the broker controller instead.

File: config/core/configmaps/data-residency.yaml (hunk 2 above)
Comment (yolocs) [reply config/core/configmaps/data-residency.yaml:18]: Good point. We probably should revisit this idea if we find that we don't need to configure other services for data residency.

**Comment by zhongduo** [comment]
Open for suggestions: I have added a default config to the data residency. However, since the data residency really is optional, I am treating empty or non key as no configuration so that it will allow all region too. The only downside I can think of is that it will not catch typos in keys like 
```
"datareside.error-key':
    - us-east-1
```

Any suggestions?

**Comment by zhongduo** [comment]
I personally think it is better to make this its own commit, and the logic to propagate to controller etc another one, as it will make the history look cleaner.

File: pkg/reconciler/intevents/topic/topic.go [hunk 4]
```diff
@@ -156,8 +160,16 @@ func (r *Reconciler) reconcileTopic(ctx context.Context, topic *v1.Topic) error
 			logging.FromContext(ctx).Desugar().Error("Topic does not exist and the topic policy doesn't allow creation")
 			return fmt.Errorf("Topic %q does not exist and the topic policy doesn't allow creation", topic.Spec.Topic)
 		} else {
+			topicConfig := &pubsub.TopicConfig{}
+			if r.dataresidencyStore != nil {
+				if dataresidencyCfg := r.dataresidencyStore.Load(); dataresidencyCfg != nil {
+					if dataresidencyCfg.DataResidencyDefaults.ComputeAllowedPersistenceRegions(topicConfig) {
+						r.Logger.Infow("Updated Topic Config", zap.Any("topicConfig", *topicConfig))
```
Comment (grantr) [inline pkg/reconciler/intevents/topic/topic.go:167]: Same comment as above: this should be Debug with a more specific message.

File: pkg/reconciler/intevents/topic/topic.go [hunk 5]
```diff
@@ -156,8 +160,16 @@ func (r *Reconciler) reconcileTopic(ctx context.Context, topic *v1.Topic) error
...
+					}
+				}
+			}
```
Comment (grantr) [inline pkg/reconciler/intevents/topic/topic.go:170]: Is there a unit test for this code? I couldn't find one.

**Review by grantr — COMMENTED** [review]
Thanks @zhongduo!

Looks like we're missing a few tests. All my other comments are optional nits.

File: pkg/apis/configs/dataresidency/defaults.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+	return d.scoped().AllowedPersistenceRegions
+}
+
+// ComputeAllowedPersistenceRegions computes the final message storage policy in
+// topicConfig. Return true if the topicConfig is updated.
+func (d *Defaults) ComputeAllowedPersistenceRegions(topicConfig *pubsub.TopicConfig) bool {
+	// We can do subset of both in the future, but for now, we just overwrite the
+	// configuration as the relationship between region and zones are not clear to handle,
+	// eg. us-east1 vs us-east1-a. Important note: setting the AllowedPersistenceRegions
+	// to empty string slice is an error, should set it to nil for all regions.
```
[the remark this answers is not in the document]
Comment (zhongduo) [reply pkg/apis/configs/dataresidency/defaults.go:57]: The current logic is to overwrite or do nothing. So either the user input or configuration map can be wrong, for example, the configuration has AllowedPersistenceRegions set to `unknown`. But I am reluctant to do validation here as proper validation is a difficult problem that require background knowledge about how GCP designs the regions and zones. The idea is that this function assumes empty or valid input and do simple calculation, errors will be detected when creating topic which will include detailed reason from GCP.

File: config/core/configmaps/data-residency.yaml [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    # options and document them in a way that is accessible
+    # to users that `kubectl edit` this config map.
+    #
+    # These sample configuration options may be copied out of
+    # this example block and unindented to be in the data block
+    # to actually change the configuration.
+
+    # default-dataresidency-config is the configuration for determining the default
+    # data residency to apply to all objects that require data residency but,
+    # do not specify it. This is expected to be Channels and Sources.
```
[the remark this answers is not in the document]
Comment (zhongduo) [reply config/core/configmaps/data-residency.yaml]: Fix as suggested.

[the remark this answers is not in the document]
Comment (zhongduo) [reply config/core/configmaps/data-residency.yaml]: Fixed as suggested.

File: pkg/apis/configs/dataresidency/dataresidency_defaults.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+	"encoding/json"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	// configName is the name of config map for the default data residency that
+	// Sources and brokers should use.
```
[the remark this answers is not in the document]
Comment (zhongduo) [reply pkg/apis/configs/dataresidency/dataresidency_defaults.go]: Fixed as suggested.

File: pkg/reconciler/broker/broker.go [hunk 9]
```diff
@@ -145,6 +149,13 @@ func (r *Reconciler) reconcileDecouplingTopicAndSubscription(ctx context.Context
 	// Check if topic exists, and if not, create it.
 	topicID := resources.GenerateDecouplingTopicName(b)
 	topicConfig := &pubsub.TopicConfig{Labels: labels}
+	if r.dataresidencyStore != nil {
+		if dataresidencyConfig := r.dataresidencyStore.Load(); dataresidencyConfig != nil {
+			if dataresidencyConfig.DataResidencyDefaults.ComputeAllowedPersistenceRegions(topicConfig) {
+				logging.FromContext(ctx).Info("Updated Topic Config for Broker", zap.Any("topicConfig", *topicConfig))
```
[the remark this answers is not in the document]
Comment (zhongduo) [reply pkg/reconciler/broker/broker.go]: Fixed as suggested.

File: pkg/reconciler/broker/broker.go [hunk 10]
```diff
@@ -145,6 +149,13 @@ func (r *Reconciler) reconcileDecouplingTopicAndSubscription(ctx context.Context
...
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (zhongduo) [reply pkg/reconciler/broker/broker.go:158]: Fixed as suggested.

File: pkg/reconciler/intevents/topic/topic.go (hunk 4 above)
Comment (zhongduo) [reply pkg/reconciler/intevents/topic/topic.go]: Fixed as suggested.

File: pkg/reconciler/trigger/trigger.go [hunk 11]
```diff
@@ -223,6 +226,13 @@ func (r *Reconciler) reconcileRetryTopicAndSubscription(ctx context.Context, tri
 	// Check if topic exists, and if not, create it.
 	topicID := resources.GenerateRetryTopicName(trig)
 	topicConfig := &pubsub.TopicConfig{Labels: labels}
+	if r.dataresidencyStore != nil {
+		if dataresidencyConfig := r.dataresidencyStore.Load(); dataresidencyConfig != nil {
+			if dataresidencyConfig.DataResidencyDefaults.ComputeAllowedPersistenceRegions(topicConfig) {
+				logging.FromContext(ctx).Info("Updated Topic Config for Trigger", zap.Any("topicConfig", *topicConfig))
```
[the remark this answers is not in the document]
Comment (zhongduo) [reply pkg/reconciler/trigger/trigger.go]: Fixed as suggested.

File: pkg/reconciler/trigger/trigger.go [hunk 12]
```diff
@@ -223,6 +226,13 @@ func (r *Reconciler) reconcileRetryTopicAndSubscription(ctx context.Context, tri
...
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (zhongduo) [reply pkg/reconciler/trigger/trigger.go:235]: Test aded, the labels is also checked.

File: pkg/reconciler/broker/broker.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (zhongduo) [reply pkg/reconciler/broker/broker.go:158]: Test aded, the labels is also checked.

File: pkg/reconciler/intevents/topic/topic.go (hunk 5 above)
Comment (grantr) [reply pkg/reconciler/intevents/topic/topic.go:170]: This code still uses gpubsub. Instead of adding tests here, @zhongduo and I agreed that we should try to switch away from gpubsub (#1739) first and follow up with tests after.

**Review by grantr — COMMENTED** [review]
Very close. I have one more testing request (test for empty AllowedPersistenceRegions) and one optional suggestion.

**Comment by grantr** [comment]
> The only downside I can think of is that it will not catch typos in keys like

This is fine for now - it's the same behavior for our other configmaps. A future improvement can add validation webhooks to verify configmap structure.

**Comment by grantr** [comment]
> it's the same behavior for our other configmaps.

Oops, I was wrong about this. We do validate the structure of other configmaps, see https://github.com/google/knative-gcp/blob/master/cmd/webhook/main.go#L171. Adding that can be a followup PR.

File: pkg/reconciler/broker/broker_test.go [hunk 13]
```diff
@@ -304,6 +308,63 @@ func TestAllCases(t *testing.T) {
...
+					Namespace: system.Namespace(),
+				},
+				Data: map[string]string{
+					// Note that the data is in yaml, so no tab is allowed, use spaces instead.
+					"default-dataresidency-config": `
+  clusterDefaults:    
+    messagestoragepolicy.allowedpersistenceregions:
+      - us-east1`,
+				},
+			},
```
[the remark this answers is not in the document]
Comment (zhongduo) [reply pkg/reconciler/broker/broker_test.go]: Fixed as suggested.

File: pkg/reconciler/broker/broker_test.go [hunk 14]
```diff
@@ -304,6 +308,63 @@ func TestAllCases(t *testing.T) {
 			TopicExists("cre-bkr_testnamespace_test-broker_abc123"),
```
[the remark this answers is not in the document]
Comment (zhongduo) [reply pkg/reconciler/broker/broker_test.go]: Fixed as suggested.

**Review by grantr — APPROVED** [review]
Thanks @zhongduo!

/lgtm
